### PR TITLE
Add `prefer-regexp-test` rule

### LINKED ISSUE #732: Rule proposal: `prefer-regexp-test`
Prefer `RegExp.prototype.test` over `String.prototype.match` for truthiness checks of (inline) regexes.

- [Performance](https://stackoverflow.com/a/10940138/271577)
- More semantically clear at a glance how the result is expected to be used. Admittedly, newcomers may think "match" sounds more like it would be a truthy check than "test", but at least for those familiar with the language details, it is more confined and clear as to intent.

## Fail

```js
if (str.match(/hasTheMagic/)) {
}
```

```js
const func = str.match(/hasTheMagic/) ? process(str) : null;
```

## Pass

```js
if ((/hasTheMagic/).test(str)) {
}
```

```js
const func = (/hasTheMagic/).test(str) ? process(str) : null;
```

```js
const matches = str.match(/hasTheMagic/);
```

(The following passes given that changing to `test` could be problematic since it could contain a global regex and such expressions passed to `test` (as with `exec`) behave non-deterministically due to `lastIndex` incrementing)
```js
if (str.match(hasTheMagicRegex)) {
}
```

Apologies this is not a "funded" request with PR planned--just if you felt it fit.

### PR DESCRIPTION
**Pull request by fisker**
Fixes #732

### CODE REVIEW

File: rules/prefer-regexp-test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	name: 'match',
+	length: 1
+});
+
+const create = context => {
+	const sourceCode = context.getSourceCode();
+
+	return {
+		[regExpExecCallSelector](node) {
+			if (isBooleanNode(node)) {
```
Comment (sindresorhus) [inline rules/prefer-regexp-test.js:29]: I would do an early-return here too.

File: rules/prefer-regexp-test.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+							regexpNode.type === 'Identifier' ||
+							regexpNode.type === 'MemberExpression' ||
+							regexpNode.type === 'CallExpression' ||
+							(regexpNode.type === 'NewExpression' && regexpText.endsWith(')'))
+						)
+					) {
+						regexpText = `(${regexpText})`;
+					}
+
+					// All nodes pass `isBooleanNode` can't have ASI problem
```
Comment (sindresorhus) [inline rules/prefer-regexp-test.js:82]:
```suggestion
					// The nodes that pass `isBooleanNode` cannot have an ASI problem.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
This is not waiting for https://github.com/sindresorhus/eslint-plugin-unicorn/pull/976, right?

**Comment by fisker** [comment]
> This is not waiting for #976, right?

Current implementation not relay on #976, but better swith to the utility I'm going to add in #976 before merge.

**Comment by sindresorhus** [comment]
> but better swith to that utility I'm going to add in #976 before merge.

👍

**Comment by sindresorhus** [comment]
Tests are failing.

**Comment by fisker** [comment]
Fixed.

**Pull request merged**